src/controllers/controller.bar.js

```diff
@@ -496,16 +495,20 @@ export default class BarController extends DatasetController {
 
 }
 
-BarController.prototype.dataElementType = Rectangle;
-
-BarController.prototype.dataElementOptions = [
-	'backgroundColor',
-	'borderColor',
-	'borderSkipped',
-	'borderWidth',
-	'barPercentage',
-	'barThickness',
-	'categoryPercentage',
-	'maxBarThickness',
-	'minBarLength'
-];
+BarController.id = 'bar';
+BarController.defaults = defaults;
+BarController.preRegister = () => {
+	BarController.prototype.dataElementType = Rectangle;
```

Comment:
I can't find anyone else online suggesting that `prototype` caused them an issue. Maybe babel is transforming it first which is causing issues? (e.g. `Object.defineProperty(BarController.prototype, ...` or something). Otherwise I think I'd report an issue to Rollup and see what they say before we introduce this workaround because it's a bit weird to me that this would cause trouble but `BarController.id = 'bar';` wouldn't

Reply:
take a look at 

https://rollupjs.org/repl/?version=2.10.8&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnRlc3QlN0QlMjBmcm9tJTIwJy4lMkZtb2RlbCclM0IlNUNuJTVDbnRlc3QoKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1vZGVsLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNsYXNzJTIwVGVzdCUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0ZXN0KCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ3Rlc3QnKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTdEJTVDbiU1Q25UZXN0LmlkJTIwJTNEJTIwJ0EnJTNCJTVDbiU1Q25leHBvcnQlMjBjbGFzcyUyMFRlc3QyJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHRlc3QoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBjb25zb2xlLmxvZygndGVzdDInKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTdEJTVDbiU1Q25UZXN0Mi5wcm90b3R5cGUuaWQlMjAlM0QlMjAnQSclM0IlNUNuJTVDbmV4cG9ydCUyMGNsYXNzJTIwVGVzdDMlMjAlN0IlNUNuJTIwJTIwJTIwJTIwdGVzdCgpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbnNvbGUubG9nKCd0ZXN0MycpJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDblRlc3QzLmElMjAlM0QlMjAlN0IlN0QlM0IlNUNuVGVzdDMuYS5pZCUyMCUzRCUyMCdBJyUzQiU1Q24lNUNuJTVDbmV4cG9ydCUyMGNsYXNzJTIwVGVzdDQlMjAlN0IlNUNuJTIwJTIwJTIwJTIwdGVzdCgpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbnNvbGUubG9nKCd0ZXN0NCcpJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDbk9iamVjdC5kZWZpbmVQcm9wZXJ0eShUZXN0NCUyQyUyMCdpZCclMkMlMjAlN0J2YWx1ZSUzQSUyMCdBJyU3RCklM0IlNUNuJTVDbmV4cG9ydCUyMGZ1bmN0aW9uJTIwdGVzdCgpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGNvbnNvbGUubG9nKCd0ZXN0RycpJTNCJTVDbiU3RCU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

and you can play around yourself. btw. webpack is not better, in this simple test it not even shaked out the `.id` variant

Reply:
Should we move `id` inside as well then to help webpack?

Reply:
then `.preRegister` still remains which has the same effect. 

I don't know how to have something like `.registerController(LineController)` without having additional attributes like id, defaults, preRegister. 

one could extract everything to a separate method: `registerLineController` but this be not as handy

Reply:
Good point. Thanks for that repl! It's very informative and helpful

I filed an [issue with rollup](https://github.com/rollup/rollup/issues/3587) to see if there's anything they can do on their side, but we should figure out what's possible just on our end in the meantime

> I don't know how to have something like .registerController(LineController) without having additional attributes like id, defaults, preRegister.

What if we do something like:
```
export function line() {
  const defaults = { ... };
  registry.registerContoller('line', LineController, defaults);
}
```

Then users could do:
```
import { Chart, line, categoryScale, lineScale } from 'chart.js';
[ line, categoryScale, lineScale ].forEach(f => f());
```